cmake/ia2.cmake

```diff
@@ -0,0 +1,272 @@
+# Creates a compartmentalized IA2 target
+#
+# add_ia2_compartment(<name> <EXECUTABLE|LIBRARY> PKEY <n> SOURCES <src>...
+#                     [INCLUDE_DIRECTORIES <dir>...] [COMPARTMENT_LIBRARIES <lib>...]
+#                     [ENABLE_UBSAN])
+#
+# Creates either an IA2 executable or shared library called <name> with the given pkey <n>.
+#
+# <name> - Target name.
+# PKEY (required) - The pkey for the compartment of this binary.
+# SOURCES (required) - The set of source files.
+# INCLUDE_DIRECTORIES (optional) - include directories used to build the executable. If
+#       a relative path this will be interpreted as relative to the current source
+#       directory.
+# COMPARTMENT_LIBRARIES (optional) - compartmentalized libraries to link against.
+# ENABLE_UBSAN - If present, enable UBSAN for this executable.
+function(add_ia2_compartment NAME TYPE)
+  # Parse options
+  set(options ENABLE_UBSAN)
+  set(oneValueArgs PKEY)
+  set(multiValueArgs LIBRARIES SOURCES INCLUDE_DIRECTORIES)
+  cmake_parse_arguments(ARG "${options}" "${oneValueArgs}"
+    "${multiValueArgs}" ${ARGN})
+
+  set(VALID_TYPES EXECUTABLE LIBRARY)
+  if(NOT "${TYPE}" IN_LIST VALID_TYPES)
+    message(FATAL_ERROR "${TYPE} is not a valid type, expected one of ${VALID_TYPES}")
+  endif()
+
+  if("${TYPE}" STREQUAL "EXECUTABLE")
+    add_executable(${NAME})
+  elseif("${ARG_PKEY}" GREATER "0")
+    set(UNPADDED_LIB ${NAME}_unpadded)
+    add_library(${UNPADDED_LIB} SHARED)
+    pad_tls_library(${UNPADDED_LIB} ${NAME})
+    set(NAME ${UNPADDED_LIB})
+  else()
+    add_library(${NAME} SHARED)
+  endif()
+
+  target_compile_definitions(${NAME} PRIVATE
+    IA2_ENABLE=1
+    PKEY=${ARG_PKEY}
+  )
+  set_target_properties(${NAME} PROPERTIES PKEY ${ARG_PKEY})
+  target_compile_options(${NAME} PRIVATE
+    "-Werror=incompatible-pointer-types"
+    "-fPIC"
+  )
+
+  set_property(TARGET ${NAME} PROPERTY ORIGINAL_SOURCES ${ARG_SOURCES})
+  set_property(TARGET ${NAME} PROPERTY ORIGINAL_INCLUDE_DIRECTORIES ${ARG_INCLUDE_DIRECTORIES})
+
+  if (ARG_ENABLE_UBSAN)
+    # UBSAN requires passing this as both a compiler and linker flag
+    target_compile_options(${NAME} PRIVATE "-fsanitize=undefined")
+    target_link_options(${NAME} PRIVATE "-fsanitize=undefined")
+  endif()
+
+  target_link_libraries(${NAME} PRIVATE dl libia2 partition-alloc)
+  target_link_options(${NAME} PRIVATE "-Wl,--export-dynamic")
```

Comment:
We previously only had this flag for the main executable, though I think for shared libs it's a no-op anyway(?), so it's not a big deal.

Reply:
Should be fine, yeah.